docs/architecture.md

```diff
@@ -0,0 +1,42 @@
+# Architecture
+
+## Introduction
+
+[osbuild-composer](https://github.com/osbuild/osbuild-composer)'s role is to be the spider in a web that allows a service like architecture
+around [osbuild](https://github.com/osbuild/osbuild) at the lower level. At a glance osbuild-composer provides an API for other tools to talk to and to distribute any
+requests to workers which execute tasks as needed.
+
+## Common
+
+### Jobqueue
+
+If an API request results in a task that needs to be executed by a worker then a job is created in the jobqueue. osbuild-composer uses PostgreSQL as the backend for its jobqueue.
+
+### Worker
+
+Workers are separate processes that look at the jobqueue and take any jobs that they can perform. Jobs are differentiated on architecture and/or operating system.
+
+#### Jobsite
+
+When a worker has accepted a job it will setup a temporary area to do this work in. This area is called the jobsite. The area takes the form of a temporary network in which a few processes are ran that are optionally ran in separate virtual machines.
+
+The jobsite goes through several distinct phases which happen in lockstep between the `manager` and the `builder`, these phases are:
+
+- Claim -- The manager ensures the builder has started up and is available.
```

Comment:
I would love if we could revisit these states at some point. I feel `/build`, `/progress`, `export` is all we need, having a single upload endpoint has the advantage that it's a single transfer, it's very simple to drive via e.g. curl and we can still do all checks and include sources by doing e.g. multipart/form-data and while streaming validating the content.